File: frontend/src/types/product.ts
import type { StatusValue } from "./common";

export interface ProductCategory {
  id: number;
  parentId: number;
  name: string;
  status: StatusValue;
  sort: number;
  children?: ProductCategory[];
  createdAt: string;
  updatedAt: string;
}

export interface ProductCategoryForm {
  parentId: number;
  name: string;
  status: StatusValue;
  sort: number;
}

export interface Brand {
  id: number;
  name: string;
  logo: string;
  status: StatusValue;
  sort: number;
  createdAt: string;
  updatedAt: string;
}

export interface BrandForm {
  name: string;
  logo: string;
  status: StatusValue;
  sort: number;
}

export type SkuSpec = Record<string, string>;

export interface ProductSku {
  id: number;
  productId: number;
  code: string;
  specs: SkuSpec;
  priceCents: number;
  stock: number;
  status: StatusValue;
  createdAt: string;
  updatedAt: string;
}

export interface ProductSkuForm {
  id?: number;
  code: string;
  specs: SkuSpec;
  priceCents: number;
  stock: number;
  status: StatusValue;
}

export interface Product {
  id: number;
  name: string;
  code: string;
  categoryId: number;
  brandId: number;
  mainImage: string;
  minPriceCents: number;
  maxPriceCents: number;
  status: StatusValue;
  description: string;
  skus: ProductSku[];
  createdAt: string;
  updatedAt: string;
}

export interface ProductForm {
  name: string;
  code: string;
  categoryId: number;
  brandId: number;
  mainImage: string;
  status: StatusValue;
  description: string;
  skus: ProductSkuForm[];
}

export interface SpecPair {
  key: string;
  value: string;
}

File: backend/internal/model/product.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type ProductCategory struct {
	ID        uint              `gorm:"primaryKey" json:"id"`
	ParentID  uint              `gorm:"not null;default:0;index" json:"parentId"`
	Name      string            `gorm:"size:64;not null;uniqueIndex" json:"name"`
	Status    int               `gorm:"not null" json:"status"`
	Sort      int               `gorm:"not null;default:0" json:"sort"`
	Children  []ProductCategory `gorm:"-" json:"children,omitempty"`
	CreatedAt time.Time         `json:"createdAt"`
	UpdatedAt time.Time         `json:"updatedAt"`
	DeletedAt gorm.DeletedAt    `gorm:"index" json:"-"`
}

type Brand struct {
	ID        uint           `gorm:"primaryKey" json:"id"`
	Name      string         `gorm:"size:64;not null;uniqueIndex" json:"name"`
	Logo      string         `gorm:"size:255" json:"logo"`
	Status    int            `gorm:"not null" json:"status"`
	Sort      int            `gorm:"not null;default:0" json:"sort"`
	CreatedAt time.Time      `json:"createdAt"`
	UpdatedAt time.Time      `json:"updatedAt"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}

type Product struct {
	ID            uint            `gorm:"primaryKey" json:"id"`
	Name          string          `gorm:"size:128;not null" json:"name"`
	Code          string          `gorm:"size:64;not null;uniqueIndex" json:"code"`
	CategoryID    uint            `gorm:"not null;index" json:"categoryId"`
	BrandID       uint            `gorm:"not null;index" json:"brandId"`
	MainImage     string          `gorm:"size:255" json:"mainImage"`
	MinPriceCents int64           `gorm:"not null;default:0" json:"minPriceCents"`
	MaxPriceCents int64           `gorm:"not null;default:0" json:"maxPriceCents"`
	Status        int             `gorm:"not null" json:"status"`
	Description   string          `gorm:"type:text" json:"description"`
	Category      ProductCategory `json:"category,omitempty"`
	Brand         Brand           `json:"brand,omitempty"`
	Skus          []ProductSKU    `json:"skus,omitempty"`
	CreatedAt     time.Time       `json:"createdAt"`
	UpdatedAt     time.Time       `json:"updatedAt"`
	DeletedAt     gorm.DeletedAt  `gorm:"index" json:"-"`
}

type ProductSKU struct {
	ID         uint           `gorm:"primaryKey" json:"id"`
	ProductID  uint           `gorm:"not null;index" json:"productId"`
	Code       string         `gorm:"size:64;not null;uniqueIndex" json:"code"`
	Specs      JSONMap        `gorm:"type:json;not null" json:"specs"`
	PriceCents int64          `gorm:"not null" json:"priceCents"`
	Stock      int            `gorm:"not null;default:0" json:"stock"`
	Status     int            `gorm:"not null" json:"status"`
	CreatedAt  time.Time      `json:"createdAt"`
	UpdatedAt  time.Time      `json:"updatedAt"`
	DeletedAt  gorm.DeletedAt `gorm:"index" json:"-"`
}

func Models() []interface{} {
	return []interface{}{
		&AdminUser{},
		&Role{},
		&Menu{},
		&APIPermission{},
		&AdminUserRole{},
		&RoleMenu{},
		&RoleAPIPermission{},
		&ProductCategory{},
		&Brand{},
		&Product{},
		&ProductSKU{},
	}
}
